File: datastruct/heap/MaxHeap.java

public class MaxHeap<E extends Comparable<E>> {

    private Array<E> data;

    public MaxHeap(int capacity){
        data = new Array<>(capacity);
    }

    public MaxHeap(){
        data = new Array<> ();
    }

    // 通过 heapify 操作将一个数组整理成最大堆
    public MaxHeap(E[] arr){

        data = new Array<> (arr);
        for(int i = getParent(arr.length - 1); i >= 0; i--){
            siftDown(i);
        }
        
    }

    // 返回堆中元素个数
    public int getSize(){
        return data.getSize();
    }

    // 堆是否为空
    public boolean isEmpty(){
        return data.isEmpty();
    }

    // 完全二叉树表示下，返回给定索引所在结点的左孩子的索引
    private int getLeftChild(int index){
        return 2 * index + 1;
    }

    // 完全二叉树表示下，返回给定索引所在结点的右孩子的索引
    private int getRightChild(int index) {
        return 2 * index + 2;
    }

    // 完全二叉树表示下，返回给定索引所咋结点的父结点的索引
    private int getParent(int index) {
        if (index == 0) {
            throw new IllegalArgumentException("index-0 doesn't have parent");
        }
        return (index - 1) / 2;
    }

    // 向堆添加一个新元素
    public void add(E e){
        data.addToLast(e);
        siftUP(data.getSize() - 1);
    }

    // 上浮操作
    private void siftUP(int k){

        while( k > 0 && data.get(getParent(k)).compareTo(data.get(k)) < 0){
            data.swap(k, getParent(k));
            k = getParent(k);
        }

    }

    // 返回堆内最大值
    public E findMax(){
        if (data.getSize() == 0) {
            throw new IllegalArgumentException("Heap is empty");
        }
        return data.get(0);
    }

    // 返回对内最大值并从堆内删除
    public E extractMax(){

        E ret = findMax();

        data.swap(0, data.getSize() - 1);
        data.removeLast();
        siftDown(0);

        return ret;
    }

    // 下沉操作
    private void siftDown(int k){

        while (getLeftChild(k) < data.getSize()) {
            int j = getLeftChild(k);
            if(j + 1 < data.getSize() && data.get(j).compareTo(data.get(j + 1)) < 0)
                j++;
            if(data.get(j).compareTo(data.get(k)) < 0)
                break;
            data.swap(k, j);
            k = j;
        }
    }

    // 取出堆中最大值，并加入一个新元素
    public E replace(E e){
        E ret = findMax();
        data.set(0, e);
        siftDown(0);
        return ret;
    }
}

File: datastruct/heap/MaxHeap.py
from Array import Array

class MaxHeap(object):

    def __init__(self, arr=None):

        if arr is None:
            self.__data = Array()
        elif isinstance(arr, int):
            self.__data = Array(arr)
        else:
            # 如果 arr 不为空，使用 heapify 操作将 arr 整理成堆
            self.__data = Array(arr)
            i = (self.get_size() - 2) // 2
            while i >= 0:
                self.__sift_down(i)
                i -= 1


    # 堆大小
    def get_size(self):
        return self.__data.get_size()

    # 堆是否为空
    def is_empty(self):
        return self.__data.is_empty()

    # 起始索引为0的完全二叉树下，返回给定索引的父结点索引
    def get_parent(self, index: int):

        if index == 0:
            raise IndexError('index-0 has no parent')
        return (index - 1) // 2

    # 起始索引为0的完全二叉树下，返回给定索引的左孩子索引
    def get_left_child(self, index: int):
        return 2 * index + 1

    # 起始索引为0的完全二叉树下，返回给定索引的右孩子索引
    def get_right_child(self, index: int):
        return 2 * index + 2

    # 向堆内添加新元素 e
    def add(self, e):
        self.__data.add_to_last(e)
        self.__sift_up(self.__data.get_size() - 1)
    
    # 上浮操作
    def __sift_up(self, index: int):

        while index > 0 and self.__data.get(self.get_parent(index)) < self.__data.get(index):
            self.__data.swap(index, self.get_parent(index))
            index = self.get_parent(index)

    # 获取堆内最大元素
    def get_max(self):
        if self.get_size() == 0:
            raise IndexError('heap is empty')
        return self.__data.get(0)

    # 获取并从堆内删除最大元素
    def extract_max(self):
        ret = self.get_max()

        self.__data.set(0, ret)
        self.__data.remove_last()
        self.__sift_down(0)

        return ret

    # 下沉操作
    def __sift_down(self, index: int):

        while self.get_left_child(index) < self.get_size():

            j = self.get_left_child(index)
            if j + 1 < self.get_size() and self.__data.get(j + 1) > self.__data.get(j):
                j += 1
            if self.__data.get(index) > self.__data.get(j):
                break
            else:
                self.__data.swap(index, j)
                index = j
    
    # 取出堆内元素最大值，并替换为新值
    def replace(self, ele):
        ret = self.get_max()
        self.__data.set(0, ele)
        self.__sift_down(0)
        return ret
